### ISSUE TITLE
EXPORT_TAGS no longer immune to once warnings

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#15688]
Migrated from [rt.perl.org#129981](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=129981) (status was 'resolved')

Searchable as RT129981$

**Comment by p5pRT** [comment]
### From @haarg

This is a bug report for perl from haarg@&#8203;haarg\.org\,
generated with the help of perlbug 1\.39 running under perl 5\.18\.2\.




---
Perl 5\.25\.6 \(specifically eafd371\) changed the code that magicalizes variables\,
which is how various variables are made immune to once warnings\.  Previously\,
any variable starting with EXPORT would be immune\.  Now\, only a specific list of
variables is immune\.  This list doesn't include EXPORT\_TAGS\.  This caused
Sub&#8203;::Exporter&#8203;::Progressive to start issuing warnings\, which is used by
Devel&#8203;::GlobalDestruction and thus Moose\, Moo\, and other modules\.

I think EXPORT\_TAGS should be added to the list of variables to be made immune\.



---


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=low


---
Site configuration information for perl 5\.18\.2&#8203;:

Configured by root at Sat Jul 30 14&#8203;:47&#8203;:24 PDT 2016\.

Summary of my perl5 \(revision 5 version 18 subversion 2\) configuration&#8203;:

&nbsp;  Platform&#8203;:
&nbsp;    osname=darwin\, osvers=16\.0\, archname=darwin\-thread\-multi\-2level
&nbsp;    uname='darwin osx202\.apple\.com 16\.0 darwin kernel version 15\.0\.0&#8203;:
wed apr 6 00&#8203;:55&#8203;:38 pdt 2016;
root&#8203;:xnu\-3247\.1\.106\.2\.8~1development\_x86\_64 x86\_64 '
&nbsp;    config\_args='\-ds \-e \-Dprefix=/usr \-Dccflags=\-g  \-pipe  \-Dldflags=
\-Dman3ext=3pm \-Duseithreads \-Duseshrplib \-Dinc\_version\_list=none
\-Dcc=cc'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-arch i386 \-arch x86\_64 \-g \-pipe \-fno\-common
\-DPERL\_DARWIN \-fno\-strict\-aliasing \-fstack\-protector'\,
&nbsp;    optimize='\-Os'\,
&nbsp;    cppflags='\-g \-pipe \-fno\-common \-DPERL\_DARWIN \-fno\-strict\-aliasing
\-fstack\-protector'
&nbsp;    ccversion=''\, gccversion='4\.2\.1 Compatible Apple LLVM 8\.0\.0
\(clang\-800\.0\.34\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc \-mmacosx\-version\-min=10\.12'\, ldflags ='\-arch i386 \-arch
x86\_64 \-fstack\-protector'
&nbsp;    libpth=/usr/lib /usr/local/lib
&nbsp;    libs=
&nbsp;    perllibs=
&nbsp;    libc=\, so=dylib\, useshrplib=true\, libperl=libperl\.dylib
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=bundle\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags='\-arch i386 \-arch x86\_64 \-bundle
\-undefined dynamic\_lookup \-fstack\-protector'

Locally applied patches&#8203;:
&nbsp;    /Library/Perl/Updates/\<version> comes before system perl directories
&nbsp;    installprivlib and installarchlib points to the Updates directory



---
@&#8203;INC for perl 5\.18\.2&#8203;:
&nbsp;    /Library/Perl/5\.18/darwin\-thread\-multi\-2level
&nbsp;    /Library/Perl/5\.18
&nbsp;    /Network/Library/Perl/5\.18/darwin\-thread\-multi\-2level
&nbsp;    /Network/Library/Perl/5\.18
&nbsp;    /Library/Perl/Updates/5\.18\.2
&nbsp;    /System/Library/Perl/5\.18/darwin\-thread\-multi\-2level
&nbsp;    /System/Library/Perl/5\.18
&nbsp;    /System/Library/Perl/Extras/5\.18/darwin\-thread\-multi\-2level
&nbsp;    /System/Library/Perl/Extras/5\.18
&nbsp;    \.



---
Environment for perl 5\.18\.2&#8203;:
&nbsp;    DYLD\_LIBRARY\_PATH \(unset\)
&nbsp;    HOME=/Users/gknop
&nbsp;    LANG=en\_US\.UTF\-8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/usr/local/sbin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @demerphq

I thought I fixed this yesterday\.

On 28 Oct 2016 05&#8203;:49\, "Graham Knop" \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> \# New Ticket Created by  Graham Knop
> \# Please include the string&#8203;:  \[perl \#129981\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=129981 >
>
>
> This is a bug report for perl from haarg@&#8203;haarg\.org\,
> generated with the help of perlbug 1\.39 running under perl 5\.18\.2\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> Perl 5\.25\.6 \(specifically eafd371\) changed the code that magicalizes
> variables\,
> which is how various variables are made immune to once warnings\.
> Previously\,
> any variable starting with EXPORT would be immune\.  Now\, only a specific
> list of
> variables is immune\.  This list doesn't include EXPORT\_TAGS\.  This caused
> Sub&#8203;::Exporter&#8203;::Progressive to start issuing warnings\, which is used by
> Devel&#8203;::GlobalDestruction and thus Moose\, Moo\, and other modules\.
>
> I think EXPORT\_TAGS should be added to the list of variables to be made
> immune\.
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>     category=core
>     severity=low
> \-\-\-
> Site configuration information for perl 5\.18\.2&#8203;:
>
> Configured by root at Sat Jul 30 14&#8203;:47&#8203;:24 PDT 2016\.
>
> Summary of my perl5 \(revision 5 version 18 subversion 2\) configuration&#8203;:
>
>   Platform&#8203;:
>     osname=darwin\, osvers=16\.0\, archname=darwin\-thread\-multi\-2level
>     uname='darwin osx202\.apple\.com 16\.0 darwin kernel version 15\.0\.0&#8203;:
> wed apr 6 00&#8203;:55&#8203;:38 pdt 2016;
> root&#8203;:xnu\-3247\.1\.106\.2\.8~1development\_x86\_64 x86\_64 '
>     config\_args='\-ds \-e \-Dprefix=/usr \-Dccflags=\-g  \-pipe  \-Dldflags=
> \-Dman3ext=3pm \-Duseithreads \-Duseshrplib \-Dinc\_version\_list=none
> \-Dcc=cc'
>     hint=recommended\, useposix=true\, d\_sigaction=define
>     useithreads=define\, usemultiplicity=define
>     useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
>     use64bitint=define\, use64bitall=define\, uselongdouble=undef
>     usemymalloc=n\, bincompat5005=undef
>   Compiler&#8203;:
>     cc='cc'\, ccflags ='\-arch i386 \-arch x86\_64 \-g \-pipe \-fno\-common
> \-DPERL\_DARWIN \-fno\-strict\-aliasing \-fstack\-protector'\,
>     optimize='\-Os'\,
>     cppflags='\-g \-pipe \-fno\-common \-DPERL\_DARWIN \-fno\-strict\-aliasing
> \-fstack\-protector'
>     ccversion=''\, gccversion='4\.2\.1 Compatible Apple LLVM 8\.0\.0
> \(clang\-800\.0\.34\)'\, gccosandvers=''
>     intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
>     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
>     ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
> lseeksize=8
>     alignbytes=8\, prototype=define
>   Linker and Libraries&#8203;:
>     ld='cc \-mmacosx\-version\-min=10\.12'\, ldflags ='\-arch i386 \-arch
> x86\_64 \-fstack\-protector'
>     libpth=/usr/lib /usr/local/lib
>     libs=
>     perllibs=
>     libc=\, so=dylib\, useshrplib=true\, libperl=libperl\.dylib
>     gnulibc\_version=''
>   Dynamic Linking&#8203;:
>     dlsrc=dl\_dlopen\.xs\, dlext=bundle\, d\_dlsymun=undef\, ccdlflags=' '
>     cccdlflags=' '\, lddlflags='\-arch i386 \-arch x86\_64 \-bundle
> \-undefined dynamic\_lookup \-fstack\-protector'
>
> Locally applied patches&#8203;:
>     /Library/Perl/Updates/\<version> comes before system perl directories
>     installprivlib and installarchlib points to the Updates directory
>
> \-\-\-
> @&#8203;INC for perl 5\.18\.2&#8203;:
>     /Library/Perl/5\.18/darwin\-thread\-multi\-2level
>     /Library/Perl/5\.18
>     /Network/Library/Perl/5\.18/darwin\-thread\-multi\-2level
>     /Network/Library/Perl/5\.18
>     /Library/Perl/Updates/5\.18\.2
>     /System/Library/Perl/5\.18/darwin\-thread\-multi\-2level
>     /System/Library/Perl/5\.18
>     /System/Library/Perl/Extras/5\.18/darwin\-thread\-multi\-2level
>     /System/Library/Perl/Extras/5\.18
>     \.
>
> \-\-\-
> Environment for perl 5\.18\.2&#8203;:
>     DYLD\_LIBRARY\_PATH \(unset\)
>     HOME=/Users/gknop
>     LANG=en\_US\.UTF\-8
>     LANGUAGE \(unset\)
>     LD\_LIBRARY\_PATH \(unset\)
>     LOGDIR \(unset\)
>     PATH=/usr/local/sbin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin
>     PERL\_BADLANG \(unset\)
>     SHELL=/bin/bash
>
>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @demerphq

Yes\. I did\.

commit 055ae706a0637a77ba440aedf596d2ebac642051
Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
Date&#8203;:   Thu Oct 27 15&#8203;:36&#8203;:27 2016 \+0200

&nbsp;    must check all the EXPORT vars in two places\!

&nbsp;    \(Silence lots of used once warnings we used to not generate\)

On 28 October 2016 at 08&#8203;:55\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:
> I thought I fixed this yesterday\.
>
>
> On 28 Oct 2016 05&#8203;:49\, "Graham Knop" \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>>
>> \# New Ticket Created by  Graham Knop
>> \# Please include the string&#8203;:  \[perl \#129981\]
>> \# in the subject line of all future correspondence about this issue\.
>> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=129981 >
>>
>>
>> This is a bug report for perl from haarg@&#8203;haarg\.org\,
>> generated with the help of perlbug 1\.39 running under perl 5\.18\.2\.
>>
>>
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> Perl 5\.25\.6 \(specifically eafd371\) changed the code that magicalizes
>> variables\,
>> which is how various variables are made immune to once warnings\.
>> Previously\,
>> any variable starting with EXPORT would be immune\.  Now\, only a specific
>> list of
>> variables is immune\.  This list doesn't include EXPORT\_TAGS\.  This caused
>> Sub&#8203;::Exporter&#8203;::Progressive to start issuing warnings\, which is used by
>> Devel&#8203;::GlobalDestruction and thus Moose\, Moo\, and other modules\.
>>
>> I think EXPORT\_TAGS should be added to the list of variables to be made
>> immune\.
>>
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> \-\-\-
>> Flags&#8203;:
>>     category=core
>>     severity=low
>> \-\-\-
>> Site configuration information for perl 5\.18\.2&#8203;:
>>
>> Configured by root at Sat Jul 30 14&#8203;:47&#8203;:24 PDT 2016\.
>>
>> Summary of my perl5 \(revision 5 version 18 subversion 2\) configuration&#8203;:
>>
>>   Platform&#8203;:
>>     osname=darwin\, osvers=16\.0\, archname=darwin\-thread\-multi\-2level
>>     uname='darwin osx202\.apple\.com 16\.0 darwin kernel version 15\.0\.0&#8203;:
>> wed apr 6 00&#8203;:55&#8203;:38 pdt 2016;
>> root&#8203;:xnu\-3247\.1\.106\.2\.8~1development\_x86\_64 x86\_64 '
>>     config\_args='\-ds \-e \-Dprefix=/usr \-Dccflags=\-g  \-pipe  \-Dldflags=
>> \-Dman3ext=3pm \-Duseithreads \-Duseshrplib \-Dinc\_version\_list=none
>> \-Dcc=cc'
>>     hint=recommended\, useposix=true\, d\_sigaction=define
>>     useithreads=define\, usemultiplicity=define
>>     useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
>>     use64bitint=define\, use64bitall=define\, uselongdouble=undef
>>     usemymalloc=n\, bincompat5005=undef
>>   Compiler&#8203;:
>>     cc='cc'\, ccflags ='\-arch i386 \-arch x86\_64 \-g \-pipe \-fno\-common
>> \-DPERL\_DARWIN \-fno\-strict\-aliasing \-fstack\-protector'\,
>>     optimize='\-Os'\,
>>     cppflags='\-g \-pipe \-fno\-common \-DPERL\_DARWIN \-fno\-strict\-aliasing
>> \-fstack\-protector'
>>     ccversion=''\, gccversion='4\.2\.1 Compatible Apple LLVM 8\.0\.0
>> \(clang\-800\.0\.34\)'\, gccosandvers=''
>>     intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
>>     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
>>     ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
>> lseeksize=8
>>     alignbytes=8\, prototype=define
>>   Linker and Libraries&#8203;:
>>     ld='cc \-mmacosx\-version\-min=10\.12'\, ldflags ='\-arch i386 \-arch
>> x86\_64 \-fstack\-protector'
>>     libpth=/usr/lib /usr/local/lib
>>     libs=
>>     perllibs=
>>     libc=\, so=dylib\, useshrplib=true\, libperl=libperl\.dylib
>>     gnulibc\_version=''
>>   Dynamic Linking&#8203;:
>>     dlsrc=dl\_dlopen\.xs\, dlext=bundle\, d\_dlsymun=undef\, ccdlflags=' '
>>     cccdlflags=' '\, lddlflags='\-arch i386 \-arch x86\_64 \-bundle
>> \-undefined dynamic\_lookup \-fstack\-protector'
>>
>> Locally applied patches&#8203;:
>>     /Library/Perl/Updates/\<version> comes before system perl directories
>>     installprivlib and installarchlib points to the Updates directory
>>
>> \-\-\-
>> @&#8203;INC for perl 5\.18\.2&#8203;:
>>     /Library/Perl/5\.18/darwin\-thread\-multi\-2level
>>     /Library/Perl/5\.18
>>     /Network/Library/Perl/5\.18/darwin\-thread\-multi\-2level
>>     /Network/Library/Perl/5\.18
>>     /Library/Perl/Updates/5\.18\.2
>>     /System/Library/Perl/5\.18/darwin\-thread\-multi\-2level
>>     /System/Library/Perl/5\.18
>>     /System/Library/Perl/Extras/5\.18/darwin\-thread\-multi\-2level
>>     /System/Library/Perl/Extras/5\.18
>>     \.
>>
>> \-\-\-
>> Environment for perl 5\.18\.2&#8203;:
>>     DYLD\_LIBRARY\_PATH \(unset\)
>>     HOME=/Users/gknop
>>     LANG=en\_US\.UTF\-8
>>     LANGUAGE \(unset\)
>>     LD\_LIBRARY\_PATH \(unset\)
>>     LOGDIR \(unset\)
>>     PATH=/usr/local/sbin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin
>>     PERL\_BADLANG \(unset\)
>>     SHELL=/bin/bash
>>
>



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @haarg

Oops\, missed that\.  This can be closed then\.

On Fri\, Oct 28\, 2016 at 2&#8203;:57 AM\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:
> Yes\. I did\.
>
> commit 055ae706a0637a77ba440aedf596d2ebac642051
> Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
> Date&#8203;:   Thu Oct 27 15&#8203;:36&#8203;:27 2016 \+0200
>
>     must check all the EXPORT vars in two places\!
>
>     \(Silence lots of used once warnings we used to not generate\)
>
> On 28 October 2016 at 08&#8203;:55\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:
>> I thought I fixed this yesterday\.
>>
>>
>> On 28 Oct 2016 05&#8203;:49\, "Graham Knop" \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>>>
>>> \# New Ticket Created by  Graham Knop
>>> \# Please include the string&#8203;:  \[perl \#129981\]
>>> \# in the subject line of all future correspondence about this issue\.
>>> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=129981 >
>>>
>>>
>>> This is a bug report for perl from haarg@&#8203;haarg\.org\,
>>> generated with the help of perlbug 1\.39 running under perl 5\.18\.2\.
>>>
>>>
>>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>>> Perl 5\.25\.6 \(specifically eafd371\) changed the code that magicalizes
>>> variables\,
>>> which is how various variables are made immune to once warnings\.
>>> Previously\,
>>> any variable starting with EXPORT would be immune\.  Now\, only a specific
>>> list of
>>> variables is immune\.  This list doesn't include EXPORT\_TAGS\.  This caused
>>> Sub&#8203;::Exporter&#8203;::Progressive to start issuing warnings\, which is used by
>>> Devel&#8203;::GlobalDestruction and thus Moose\, Moo\, and other modules\.
>>>
>>> I think EXPORT\_TAGS should be added to the list of variables to be made
>>> immune\.
>>>
>>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>>> \-\-\-
>>> Flags&#8203;:
>>>     category=core
>>>     severity=low
>>> \-\-\-
>>> Site configuration information for perl 5\.18\.2&#8203;:
>>>
>>> Configured by root at Sat Jul 30 14&#8203;:47&#8203;:24 PDT 2016\.
>>>
>>> Summary of my perl5 \(revision 5 version 18 subversion 2\) configuration&#8203;:
>>>
>>>   Platform&#8203;:
>>>     osname=darwin\, osvers=16\.0\, archname=darwin\-thread\-multi\-2level
>>>     uname='darwin osx202\.apple\.com 16\.0 darwin kernel version 15\.0\.0&#8203;:
>>> wed apr 6 00&#8203;:55&#8203;:38 pdt 2016;
>>> root&#8203;:xnu\-3247\.1\.106\.2\.8~1development\_x86\_64 x86\_64 '
>>>     config\_args='\-ds \-e \-Dprefix=/usr \-Dccflags=\-g  \-pipe  \-Dldflags=
>>> \-Dman3ext=3pm \-Duseithreads \-Duseshrplib \-Dinc\_version\_list=none
>>> \-Dcc=cc'
>>>     hint=recommended\, useposix=true\, d\_sigaction=define
>>>     useithreads=define\, usemultiplicity=define
>>>     useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
>>>     use64bitint=define\, use64bitall=define\, uselongdouble=undef
>>>     usemymalloc=n\, bincompat5005=undef
>>>   Compiler&#8203;:
>>>     cc='cc'\, ccflags ='\-arch i386 \-arch x86\_64 \-g \-pipe \-fno\-common
>>> \-DPERL\_DARWIN \-fno\-strict\-aliasing \-fstack\-protector'\,
>>>     optimize='\-Os'\,
>>>     cppflags='\-g \-pipe \-fno\-common \-DPERL\_DARWIN \-fno\-strict\-aliasing
>>> \-fstack\-protector'
>>>     ccversion=''\, gccversion='4\.2\.1 Compatible Apple LLVM 8\.0\.0
>>> \(clang\-800\.0\.34\)'\, gccosandvers=''
>>>     intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
>>>     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
>>>     ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
>>> lseeksize=8
>>>     alignbytes=8\, prototype=define
>>>   Linker and Libraries&#8203;:
>>>     ld='cc \-mmacosx\-version\-min=10\.12'\, ldflags ='\-arch i386 \-arch
>>> x86\_64 \-fstack\-protector'
>>>     libpth=/usr/lib /usr/local/lib
>>>     libs=
>>>     perllibs=
>>>     libc=\, so=dylib\, useshrplib=true\, libperl=libperl\.dylib
>>>     gnulibc\_version=''
>>>   Dynamic Linking&#8203;:
>>>     dlsrc=dl\_dlopen\.xs\, dlext=bundle\, d\_dlsymun=undef\, ccdlflags=' '
>>>     cccdlflags=' '\, lddlflags='\-arch i386 \-arch x86\_64 \-bundle
>>> \-undefined dynamic\_lookup \-fstack\-protector'
>>>
>>> Locally applied patches&#8203;:
>>>     /Library/Perl/Updates/\<version> comes before system perl directories
>>>     installprivlib and installarchlib points to the Updates directory
>>>
>>> \-\-\-
>>> @&#8203;INC for perl 5\.18\.2&#8203;:
>>>     /Library/Perl/5\.18/darwin\-thread\-multi\-2level
>>>     /Library/Perl/5\.18
>>>     /Network/Library/Perl/5\.18/darwin\-thread\-multi\-2level
>>>     /Network/Library/Perl/5\.18
>>>     /Library/Perl/Updates/5\.18\.2
>>>     /System/Library/Perl/5\.18/darwin\-thread\-multi\-2level
>>>     /System/Library/Perl/5\.18
>>>     /System/Library/Perl/Extras/5\.18/darwin\-thread\-multi\-2level
>>>     /System/Library/Perl/Extras/5\.18
>>>     \.
>>>
>>> \-\-\-
>>> Environment for perl 5\.18\.2&#8203;:
>>>     DYLD\_LIBRARY\_PATH \(unset\)
>>>     HOME=/Users/gknop
>>>     LANG=en\_US\.UTF\-8
>>>     LANGUAGE \(unset\)
>>>     LD\_LIBRARY\_PATH \(unset\)
>>>     LOGDIR \(unset\)
>>>     PATH=/usr/local/sbin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin
>>>     PERL\_BADLANG \(unset\)
>>>     SHELL=/bin/bash
>>>
>>
>
>
>
> \-\-
> perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Oct 27 23&#8203;:58&#8203;:15 2016\, demerphq wrote&#8203;:
> Yes\. I did\.
> 
> commit 055ae706a0637a77ba440aedf596d2ebac642051
> Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
> Date&#8203;:   Thu Oct 27 15&#8203;:36&#8203;:27 2016 \+0200
> 
>     must check all the EXPORT vars in two places\!
> 
>     \(Silence lots of used once warnings we used to not generate\)

Your original commit message said&#8203;:

> Another issue is that some of the code was inadverdantly
> doing prefix matches\, and not exact matchs for certain
> magic var names\. This is fixed as part of these changes\.

I do not think it was inadvertent\.  I believe there are other EXPORT\_\* variables also used\, but grep\.cpan\.me is down right now\.

In any case\, I think there \*should\* be a prefix match for EXPORT\*\.


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @demerphq

On 29 October 2016 at 16&#8203;:30\, Father Chrysostomos via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Thu Oct 27 23&#8203;:58&#8203;:15 2016\, demerphq wrote&#8203;:
>> Yes\. I did\.
>>
>> commit 055ae706a0637a77ba440aedf596d2ebac642051
>> Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
>> Date&#8203;:   Thu Oct 27 15&#8203;:36&#8203;:27 2016 \+0200
>>
>>     must check all the EXPORT vars in two places\!
>>
>>     \(Silence lots of used once warnings we used to not generate\)
>
> Your original commit message said&#8203;:
>
>> Another issue is that some of the code was inadverdantly
>> doing prefix matches\, and not exact matchs for certain
>> magic var names\. This is fixed as part of these changes\.
>
> I do not think it was inadvertent\.  I believe there are other EXPORT\_\* variables also used\, but grep\.cpan\.me is down right now\.
>
> In any case\, I think there \*should\* be a prefix match for EXPORT\*\.

We don't document such an exemption\. And when in place it means that
if people typo and EXPORT related var name then there are no warnings\.
Eg\, these should trigger used only once warnings&#8203;:

@&#8203;EXPORT\_oK= qw\( \.\.\. \);
@&#8203;EXPORTOK= qw\(\);
@&#8203;EXPORT\_\_OK=qw\(\);

But without my patch none of them will\. I am fine if we add more
specific var names\, but I really think we should have to enumerate
them\, especially as they should be documented in PERLVAR too\.

Yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @xsawyerx



On 10/29/2016 07&#8203;:47 PM\, demerphq wrote&#8203;:
> On 29 October 2016 at 16&#8203;:30\, Father Chrysostomos via RT
> \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>> On Thu Oct 27 23&#8203;:58&#8203;:15 2016\, demerphq wrote&#8203;:
>>> Yes\. I did\.
>>>
>>> commit 055ae706a0637a77ba440aedf596d2ebac642051
>>> Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
>>> Date&#8203;:   Thu Oct 27 15&#8203;:36&#8203;:27 2016 \+0200
>>>
>>>     must check all the EXPORT vars in two places\!
>>>
>>>     \(Silence lots of used once warnings we used to not generate\)
>> Your original commit message said&#8203;:
>>
>>> Another issue is that some of the code was inadverdantly
>>> doing prefix matches\, and not exact matchs for certain
>>> magic var names\. This is fixed as part of these changes\.
>> I do not think it was inadvertent\.  I believe there are other EXPORT\_\* variables also used\, but grep\.cpan\.me is down right now\.
>>
>> In any case\, I think there \*should\* be a prefix match for EXPORT\*\.
> We don't document such an exemption\. And when in place it means that
> if people typo and EXPORT related var name then there are no warnings\.
> Eg\, these should trigger used only once warnings&#8203;:
>
> @&#8203;EXPORT\_oK= qw\( \.\.\. \);
> @&#8203;EXPORTOK= qw\(\);
> @&#8203;EXPORT\_\_OK=qw\(\);
>
> But without my patch none of them will\. I am fine if we add more
> specific var names\, but I really think we should have to enumerate
> them\, especially as they should be documented in PERLVAR too\.

I don't think we should have a pattern of ^EXPORT\_\, but specify the
specific variables allowed\. Typos are just one reason\, but forced future
compatibility due to existing usage is another\. We should support what
we know and know what we support\.
